File: nextjs/src/lib/api/files.ts
/**
 * dbt-runner API: File operations
 */

import { apiClient } from './client';

// Types
export interface FileNode {
    name: string;
    path: string;
    type: 'file' | 'folder';
    size?: number;
}

export interface FileListResponse {
    path: string;
    items: FileNode[];
}

export interface FileContentResponse {
    path: string;
    content: string;
}

export interface FileSaveResponse {
    success: boolean;
    path: string;
    message?: string;
}

export interface FileCreateRequest {
    path: string;
    file_type: 'file' | 'directory';
    content?: string;
}

export interface FileCreateResponse {
    success: boolean;
    message: string;
    path: string;
}

export interface FileDeleteResponse {
    success: boolean;
    message: string;
}

export interface FileRenameResponse {
    success: boolean;
    message: string;
    old_path: string;
    new_path: string;
}

export interface FileSearchResponse {
    query: string;
    results: FileNode[];
}

export interface FileMoveResponse {
    success: boolean;
    message: string;
    source_path: string;
    dest_path: string;
}

export interface FileCopyResponse {
    success: boolean;
    message: string;
    source_path: string;
    dest_path: string;
}

export interface FileDuplicateResponse {
    success: boolean;
    message: string;
    original_path: string;
    new_path: string;
}

export interface ProjectStatusResponse {
    project_id: string;
    exists: boolean;
    has_git: boolean;
    has_dbt_project: boolean;
    has_uncommitted_changes: boolean;
    file_count: number;
    path: string;
}

// API Functions
export const filesApi = {
    /**
     * List files in a project directory
     */
    list: (projectId: string, path: string = '') => {
        const queryPath = path ? `?path=${encodeURIComponent(path)}` : '';
        return apiClient.get<FileListResponse>(`/files/${projectId}${queryPath}`);
    },

    /**
     * Read file content
     */
    read: (projectId: string, path: string) =>
        apiClient.get<FileContentResponse>(
            `/files/${projectId}/content?path=${encodeURIComponent(path)}`
        ),

    /**
     * Save file content
     */
    save: (projectId: string, path: string, content: string) =>
        apiClient.post<FileSaveResponse>(`/files/${projectId}/content`, {
            path,
            content,
        }),

    /**
     * Create a new file or directory
     */
    create: (projectId: string, request: FileCreateRequest) =>
        apiClient.post<FileCreateResponse>(`/files/${projectId}/create`, request),

    /**
     * Delete a file or directory
     */
    delete: (projectId: string, path: string) =>
        apiClient.delete<FileDeleteResponse>(
            `/files/${projectId}?path=${encodeURIComponent(path)}`
        ),

    /**
     * Rename a file or directory
     */
    rename: (projectId: string, oldPath: string, newPath: string) =>
        apiClient.put<FileRenameResponse>(
            `/files/${projectId}/rename?old_path=${encodeURIComponent(oldPath)}&new_path=${encodeURIComponent(newPath)}`
        ),

    /**
     * Search for files in a project
     */
    search: (projectId: string, query: string) =>
        apiClient.get<FileSearchResponse>(
            `/files/${projectId}/search?query=${encodeURIComponent(query)}`
        ),

    /**
     * Move a file or directory to a new location
     */
    move: (projectId: string, sourcePath: string, destPath: string) =>
        apiClient.post<FileMoveResponse>(
            `/files/${projectId}/move?source_path=${encodeURIComponent(sourcePath)}&dest_path=${encodeURIComponent(destPath)}`
        ),

    /**
     * Copy a file or directory to a new location
     */
    copy: (projectId: string, sourcePath: string, destPath: string) =>
        apiClient.post<FileCopyResponse>(
            `/files/${projectId}/copy?source_path=${encodeURIComponent(sourcePath)}&dest_path=${encodeURIComponent(destPath)}`
        ),

    /**
     * Duplicate a file or directory in the same location
     */
    duplicate: (projectId: string, path: string) =>
        apiClient.post<FileDuplicateResponse>(
            `/files/${projectId}/duplicate?path=${encodeURIComponent(path)}`
        ),

    /**
     * Get project status - check if project exists and its state
     */
    getStatus: (projectId: string) =>
        apiClient.get<ProjectStatusResponse>(`/files/${projectId}/status`),
};

File: dbt-runner/app/routers/files.py
"""
File operations router.
"""

import logging

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.auth import require_user, resolve_user_id, verify_project_ownership
from app.core.db import get_session
from app.core.dependencies import get_file_service
from app.exceptions import (
    FileNotFoundException,
    InvalidPathException,
    ProjectNotFoundException,
)
from app.models.file import FileCreateRequest, FileSaveRequest
from app.services.file_service import FileService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/files", tags=["Files"])


# Shared with every other router - see app/core/auth.py
_verify_project_ownership = verify_project_ownership


@router.get("/{project_id}")
async def list_files(
    project_id: str,
    path: str = "",
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """List files in a project directory."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.list_files(project_id, path)
    except (ProjectNotFoundException, FileNotFoundException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.get("/{project_id}/search")
async def search_files(
    project_id: str,
    query: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Search for files matching a query in the project."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.search_files(project_id, query)
    except ProjectNotFoundException as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.get("/{project_id}/content")
async def read_file(
    project_id: str,
    path: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Read file content."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.read_file(project_id, path)
    except (ProjectNotFoundException, FileNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.put("/{project_id}/content")
async def write_file(
    project_id: str,
    path: str,
    content: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Write file content."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.write_file(project_id, path, content)
    except (ProjectNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.post("/{project_id}/create")
async def create_file(
    project_id: str,
    request: FileCreateRequest,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Create a new file or directory in the project."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.create(project_id, request)
    except (ProjectNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.post("/{project_id}/content")
async def save_file_content(
    project_id: str,
    request: FileSaveRequest,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Save content to an existing file."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.save_content(project_id, request)
    except (ProjectNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.delete("/{project_id}")
async def delete_file(
    project_id: str,
    path: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Delete a file or directory from the project."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.delete(project_id, path)
    except (ProjectNotFoundException, FileNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.put("/{project_id}/rename")
async def rename_file(
    project_id: str,
    old_path: str,
    new_path: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Rename a file or directory."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.rename(project_id, old_path, new_path)
    except (ProjectNotFoundException, FileNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.post("/{project_id}/move")
async def move_file(
    project_id: str,
    source_path: str,
    dest_path: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Move a file or directory to a new location."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.move(project_id, source_path, dest_path)
    except (ProjectNotFoundException, FileNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.post("/{project_id}/copy")
async def copy_file(
    project_id: str,
    source_path: str,
    dest_path: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Copy a file or directory to a new location."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.copy(project_id, source_path, dest_path)
    except (ProjectNotFoundException, FileNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.post("/{project_id}/duplicate")
async def duplicate_file(
    project_id: str,
    path: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """Duplicate a file or directory with _copy suffix."""
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)
    try:
        return await service.duplicate(project_id, path)
    except (ProjectNotFoundException, FileNotFoundException, InvalidPathException) as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.get("/{project_id}/status")
async def get_project_status(
    project_id: str,
    claims: dict = Depends(require_user),
    session: AsyncSession = Depends(get_session),
    service: FileService = Depends(get_file_service),
):
    """
    Check project status - whether it exists and is accessible.

    Returns:
        - exists: Whether project directory exists
        - has_git: Whether project has .git directory
        - has_dbt_project: Whether project has dbt_project.yml
        - has_uncommitted_changes: Whether there are uncommitted Git changes
        - file_count: Number of files in project root
    """
    user_id = await resolve_user_id(session, claims.get("sub"), claims.get("email"))
    await _verify_project_ownership(session, project_id, user_id)

    from pathlib import Path

    from app.services.project import ProjectService

    project_service = ProjectService()
    project_path = project_service.get_path(project_id)

    result = {
        "project_id": project_id,
        "exists": False,
        "has_git": False,
        "has_dbt_project": False,
        "has_uncommitted_changes": False,
        "file_count": 0,
        "path": str(project_path),
    }

    if not project_path.exists():
        return result

    result["exists"] = True
    result["has_git"] = (project_path / ".git").exists()
    result["has_dbt_project"] = (project_path / "dbt_project.yml").exists()

    try:
        result["file_count"] = len(list(project_path.iterdir()))
    except Exception:
        pass

    if result["has_git"]:
        try:
            import subprocess

            proc = subprocess.run(
                ["git", "status", "--porcelain"],
                cwd=project_path,
                capture_output=True,
                text=True,
                timeout=5,
            )
            result["has_uncommitted_changes"] = bool(proc.stdout.strip())
        except Exception:
            pass

    return result
